spec/controllers/admin_censor_rule_controller_spec.rb

```diff
@@ -835,6 +835,16 @@ def create_censor_rule
           expect(flash[:notice]).to eq(msg)
         end
 
+        it 'purges the cache for the info request' do
+          public_body = double("public_body")
+          allow(CensorRule).to receive(:find) { @censor_rule }
+          allow(@censor_rule).to receive(:public_body) { public_body }
+          expect(public_body).to receive(:expire_requests)
+
+          put :update, :id => @censor_rule.id,
+            :censor_rule => { :text => 'different text' }
```

Comment:
I appreciate this is an existing pattern, but the level of stubbing here makes me uncomfortable.

I don't think the controller should have the logic of what to expire.  I think the spec should be this:

```ruby
it 'purges the cache for the info request' do
  expect(@censor_rule).to receive(:expire_requests)
  put :update, :id => @censor_rule.id,
               :censor_rule => { :text => 'different text' }
end
```
I know that `CensorRule#expire_requests` doesn't yet exist; I think it should.

Here's what I think we want:

```ruby
class CensorRule #…
  def expire_requests
    if info_request
      info_request.expire
    elsif user
      user.expire_requests
    elsif public_body
      public_body.expire_requests
    end
  end
 end

class AdminCensorRuleController #…
   def expire_requests_and_redirect
     @censor_rule.expire_requests

     if @censor_rule.info_request
       redirect_to admin_request_url(@censor_rule.info_request)
     elsif @censor_rule.user
       redirect_to admin_user_url(@censor_rule.user)
     elsif @censor_rule.public_body
       redirect_to admin_body_url(@censor_rule.public_body)
     else
       redirect_to admin_censor_rules_path
     end
   end
end
```

- `InfoRequest#expire` should be robustly tested
- `PublicBody#expire_requests` and `User#expire_requests` can just expect that `#expire` is called on the correct association(/s)
- `CensorRule#expire_requests` can check that `#expire_requests` is called on the correct associations
- The controller spec here can just check that `#expire_requests` is called on the `CensorRule`
- Integration tests can do a full-stack test to ensure that the end-to-end process results in the correct redactions.